### PR TITLE
rustdoc: re-elide cross-crate default trait-object lifetime bounds

### PR DESCRIPTION
**Pull request by fmease**
Hide trait-object lifetime bounds (re-exported from an external crate) if they coincide with [their default](https://doc.rust-lang.org/reference/lifetime-elision.html#default-trait-object-lifetimes).
Partially addresses #44306. Follow-up to #103885. [Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/clean_middle_ty.3A.20I.20need.20to.20add.20a.20parameter/near/307143097).

Most notably, if `std` exported something from `core` containing a type like `Box<dyn Fn()>`, then it would now be rendered as `Box<dyn Fn(), Global>` instead of `Box<dyn Fn() + 'static, Global>` (hiding `+ 'static` as it is the default in this case). Showing `Global` here is a separate issue, #80379, which is on my agenda.

Note that I am not really fond of the fact that I had to add a parameter to such a widely used function (30+ call sites) to address such a niche bug.

CC @GuillaumeGomez
Requesting a review from a compiler contributor or team member as recommended on Zulip.
r? compiler

---

@rustbot label T-compiler T-rustdoc A-cross-crate-reexports

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
The changes look good to me, thanks for working on this! I think it'd be nice to have someone from the compiler team to also take a look at this to confirm that it's the right solution and also another person from rustdoc to ensure I didn't miss something.

cc @davidtwco @notriddle

**Comment by fmease** [comment]
Spurious network failure as far as I can tell :disappointed:, could you please retry?

**Comment by davidtwco** [comment]
r? compiler

**Comment by fmease** [comment]
Let me look into the regressions… :)

**Comment by fmease** [comment]
Okay, for now I've only *rebased* my PR without any significant changes. As you can see from the perf results above, the numbers are *wild* for `cranelift-codegen` and `keccak`. These are false positives. #108815 fixed these performance ‘bimodalities’ (see also [*`keccak` bimodility* (Zulip)](https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/keccak.20bimodality)). E.g. in #108347, perf came back neutral after #108815.

Could you retry the perf (once CI is green)? That'd be splendid! Thanks in advance. Sorry it's been a while. I'm mostly occupied with other parts of rustc ^^'.

**Comment by fmease** [comment]
Small secondary regression but no actions needed (also, this is still marked as a draft). I'm sure I can improve this a bit later. First however I'd like to get some rough non-T-rustdoc feedback. Just to check if there are some obvious flaws or not :) Thanks!

@rustbot ready

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks! It'd be nice to have at nice one other review from the rustdoc team considering how big this is. Also, do you need someone from the compiler team to take a another look too?

**Comment by fmease** [comment]
> the compiler team to take a another look too?

Yes, definitely. That's why I wrote “non-T-rustdoc” ;-) @TaKO8Ki

**Comment by fmease** [comment]
Sorry, I didn't notice my PR became unmergeable. I've now rebased.

**Comment by GuillaumeGomez** [comment]
r? rust-lang/compiler

**Comment by fmease** [comment]
davidtwco has already rerolled on this PR once
r? compiler

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1883]: This indexing dance is ok, that's what `rustc_hir_analysis::collect::resolve_bound_vars::visit_segment_args` does.

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1884]: Could you add a binding for this index?

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1722]: Expression context means you are in a function body, or a static/const rhs.

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1760]: This should be `Lifetime(region)`.

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1741]: I suggest doing it in the opposite order:
- compute what the implicit lifetime is from `ObjectLifetimeDefault`,
- compare it with `region`.

File: src/librustdoc/clean/mod.rs
Comment (cjgillot) [inline src/librustdoc/clean/mod.rs:1806]: Do you have a way of using `rustc_trait_selection::traits::wf::object_region_bounds` like astconv does?

**Comment by fmease** [comment]
I'm sorry for the delay. I've greatly simplified the logic, I now handle higher-ranked regions properly (except for shadowing) and make use of `wf::object_region_bounds` (thanks for suggestion, @cjgillot!). I'm inclined to leave the `FIXME(fmease)`s regarding shadowing for a follow-up PR since they're such minute edge cases & since this PR has been open for way to long. I hope this is fine.

NB: I'm not going to properly handle object lifetime defaults inside *expression contexts* since afaik this case isn't reachable in rustdoc (we don't render fn bodies or complex const exprs at the time of this writing). I've left an anonymous `FIXME`. Passing this information down would require too many changes for no benefit.

@rustbot ready

**Comment by fmease** [comment]
Also let me know if I should weed out some comments, the code/comment ratio has become quite extreme by now after the great simplifications.

**Review by GuillaumeGomez — APPROVED** [review]

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
I personally really appreciate the code comments, they're really helpful, especially in this case. Like I said, considering it's a big PR, I'll wait for someone else from the team to also review it to be sure nothing was missed. cc @notriddle

**Review by notriddle — APPROVED** [review]
This all looks okay to me as well.

**Comment by GuillaumeGomez** [comment]
Then let's go! Thanks to everyone involved!

@bors r=notriddle,cgillot,GuillaumeGomez

**Comment by fmease** [comment]
> <kbd>perf-regression</kbd>

Right, that was to be expected (doing more work for every single trait object type) and we've already seen this in the 2nd of 3 perf runs. Let me know if this is acceptable or if I should try to make this faster (fast path or cache maybe?). I'm already delaying as much work as possible (hence `ContainerTy` instead of eagerly computing things).

**Comment by fmease** [comment]
Ok, then
@rustbot label perf-regression-triaged

**Comment by QuineDot** [comment]
Heads up: This PR seems to be based around the reference's section on default trait object lifetimes, but that section of the reference [is inaccurate in many ways.](https://github.com/rust-lang/reference/issues/1407)  I imagine the main cases for Rustdoc would be
- Trait bounds usually override type bounds, not the other way around
    ```rust
    use std::any::Any;
    fn common_example(r: &dyn Any) -> &(dyn Any + 'static) {
        // This line fails so the reference lifetime is not the default lifetime
        // (And if it was, the reference lifetime would have to be `'static`)
        let _: &'static dyn Any = r;
        r
    }
    ```
    ```rust
    trait Bound<'a, 'b>: 'a {}
    fn another_example<'a: 'b, 'b>(r: &dyn Bound<'a, 'b>) {
        // This compiles so the reference lifetime is not the default lifetime
        let _: &(dyn Bound<'a, 'b> + 'a) = r;
    }
    ```
- Type aliases bounds (or lack of bounds) take precedence over bounds on what they alias
    ```rust
    use std::fmt::Display;
    type Alias<'a, T> = &'a T;
    type Surprise<'a> = Alias<'a, dyn Display>;

    // This compiles
    fn example(s: Surprise<'_>) -> &(dyn Display + 'static) { s }
    ```

Unfortunately the exception to trait bounds overriding types bound is "in function signatures, sometimes" (when no bound parameters are late bound or `'static`).  The exact rules for what the actual default lifetime is in function signatures is [sadly a giant mess.](https://github.com/rust-lang/rust/issues/47078#issuecomment-1636690658)

I'm not sure what the best way to actually test Rustdoc on this matter is, but would be willing to give it a shot if there's some guidance on how to test Rustdoc somewhere.

**Comment by fmease** [comment]
I've been meaning to reply to [this comment of yours](https://github.com/rust-lang/rust/issues/47078#issuecomment-1636690658) for a while now but I've since forgotten about it. I'm not an expert on the details of region checking but I'm gonna check each snippet you posted over at rust-lang/reference. Some things definitely look fishy

**Comment by fmease** [comment]
Also note that most of the things you listed here and over at rust-lang/reference shouldn't be relevant to rustdoc since they only affect types found in bodies (function, closure, constant) which rustdoc doesn't (really) render.

**Pull request merged**
